src/blueapi/config.py

```diff
@@ -36,3 +40,68 @@ class ApplicationConfig(BlueapiBaseModel):
     stomp: StompConfig = Field(default_factory=StompConfig)
     env: EnvironmentConfig = Field(default_factory=EnvironmentConfig)
     logging: LoggingConfig = Field(default_factory=LoggingConfig)
+
+
+C = TypeVar("C", bound=BaseModel)
+
+
+class ConfigLoader(Generic[C]):
+    """
+    Small utility class for loading config from various sources.
+    You must define a config schema as a dataclass (or series of
+    nested dataclasses) that can then be loaded from some combination
+    of default values, dictionaries, YAML/JSON files etc.
+    """
+
+    _schema: Type[C]
+    _values: Mapping[str, Any]
+
+    def __init__(self, schema: Type[C]) -> None:
+        self._schema = schema
+        self._values = {}
+
+    def use_values(self, values: Mapping[str, Any]) -> None:
+        """
+        Use all values provided in the config, override any defaults
+        and values set by previous calls into this class.
+
+        Args:
+            values (Mapping[str, Any]): Dictionary of override values,
+                                        does not need to be exaustive
+                                        if defaults provided.
+        """
+
+        self._values = {**self._values, **values}
+
+    def load_from_yaml(self, path: Path) -> C:
+        """
+        Use all values provided in a YAML/JSON file in the
+        config, override any defaults and values set by
+        previous calls into this class.
+
+        Args:
+            path (Path): Path to YAML/JSON file
+        """
+
+        with path.open("r") as stream:
+            values = yaml.load(stream, yaml.Loader)
+        self.use_values(values)
+
+        return self.load()
+
+    def load(self) -> C:
+        """
+        Finalize and load the config as an instance of the `schema`
+        dataclass.
+
+        Returns:
+            C: Dataclass instance holding config
+        """
+
+        try:
+            return parse_obj_as(self._schema, self._values)
+        except ValidationError:
+            raise InvalidConfigError(
```

Comment:
I would allow the ValidationError to propagate, it's already indicative of what the error is, and allows you to get rid of InvalidConfigError

Reply:
I think the InvalidConfigError is a bit more descriptive, it states the expected schema for example and says that the config file is wrong. The Validation Error is also useful, and still included in the traceback when this is raised.

Reply:
I thought that would require `from`?

Reply:
Yeah, it seems not to need it. I can write a test for this maybe...